source/Ship.cpp

```diff
@@ -2645,12 +2739,158 @@ bool Ship::IsParalyzed() const
-	return (hull < minimumHull || (!crew && needsCrew));
+// If a ship is nearly out of battery energy and is slow to charge or no charge,
+// then consider it low on energy.
+bool Ship::IsEnergyLow() const
+{
+	double frames = 60.;
+	double idleEnergy = frames * GetIdleEnergyPerFrame();
+	double maxEnergy = Attributes().Get("energy capacity");
+	double totalConsumption = frames * GetEnergyConsumptionPerFrame();
+	double currentEnergy = (CanBeCarried() && GetSystem() && !landingPlanet) ? GetCurrentEnergy() : maxEnergy * .75;
+	//double currentEnergy = GetCurrentEnergy();
+	if(totalConsumption < 0.)
+	{
+		double secondsToEmpty = currentEnergy / (-totalConsumption);
+		// how quickly can it charge under no energy load
+		double secondsToFullCharge = (idleEnergy > 0.) ? (maxEnergy - currentEnergy) / idleEnergy : minimumOperatingTime + 1.;
+		if((secondsToEmpty < minimumOperatingTime && secondsToFullCharge > minimumOperatingTime) || (secondsToEmpty < minimumOperatingTime && idleEnergy <= 0.))
+			return true;
+	}
+	return false;
+}
+
+
+
+// Check if escort fleet is full of fuel.  In this case, the fighters can help
+// replenish the Carrier fuel tanker.
+bool Ship::IsEscortsFullOfFuel() const
+{
+	if(!IsYours())
+		return false;
+	return isEscortsFullOfFuel;
+}
+
+
+
+// Out of energy ships will be disabled.  Out of energy fighters will be
+// disabled if far away from the parent.  This is so the parent can board the
+// fighter however the fighter needs to not be disabled in order to board the
+// carrier.
+bool Ship::IsOutOfEnergy() const
+{
+	bool closeToParent = false;
+	bool outOfEnergy = GetCurrentEnergy() <= 0. && GetMovingEnergyPerFrame() > 0.  && GetIdleEnergyPerFrame() <= 0.;
+	if(canBeCarried && outOfEnergy)
+	{
+		if(GetParent())
+		{
+			Point dp = GetParent().get()->Position() - position;
+			Point dv = GetParent().get()->Velocity() - velocity;
+			double distanceFromParent = dp.Length();
+			double speedRelativeToParent = dv.Length();
+			closeToParent = distanceFromParent < 50. && speedRelativeToParent < 1.;
+			// Set closeToParent to false if this ship is disabled due to hull.  It requires additional assistance.
+			closeToParent &= hull >= minimumHull;
+			closeToParent &= static_cast<bool>(GetSystem());
+			// CanCarry is expensive performance-wise so should only be called if necessary.
+			if(closeToParent)
+				closeToParent &= GetParent()->BaysFree(attributes.Category());
+			// Clear the parent boarding target when fighter boards.
+			if(closeToParent && GetParent().get()->GetTargetShip().get() == this)
+				GetParent().get()->SetTargetShip(nullptr);
+		}
+		else
+			closeToParent = false;
+	}
+	// Return will always be false if not a fighter.
+	return !landingPlanet && !hasBoarded && !closeToParent && outOfEnergy;
+}
+
+
+
+// Check if ship fuel is low or check destination ship fuel (compareTo) can refuel.
+bool Ship::IsFuelLow() const
+{
+	return IsFuelLow(attributes.Get("fuel capacity"));
```

Comment:
It annoys me that we query this value twice in this case, remember this isnt java so we can have default values if we want them (and if we don't have any then we'll query fuel capacity but only once)